Split change bundles with excessive asset counts

jonathanknowles opened the pull request:
# Issue Number

ADP-727

# Overview

This PR has two main areas:

## Refactoring
- [x] Relocates all existing type-specific `equipartition` functions into the modules that define their types. We now have:
    | Module | Function |
    | -- | -- |
    | `Numeric.Util` | `equipartitionNatural` |
    | `Coin` | `equipartition` |
    | `TokenQuantity` | `equipartition` |
    | `TokenMap` | `equipartitionQuantities` |
    | `TokenMap` | `equipartitionQuantitiesWithUpperBound` |
    | `TokenBundle` | `equipartitionQuantitiesWithUpperBound` |
    
    **Justification:**
    - This PR will add **_more_** functions for splitting up bundles and maps, and the existing naming scheme was becoming too cumbersome to extend.
    - Equipartitioning a value of type `T` is something that relates to the definition of type `T`, so the natural home for such a function is within module `T`.
    
- [x] Adjusts `makeChange` to accept a record type, and unifies that type with `MakeChangeData` from the test suite.

    **Justification:**
    - Adding further arguments to this function is now less cumbersome, since the compiler will tell us that **_a field is missing_** in all the required call sites (both in implementation code and in test code), rather than producing a series of long type errors.
    - We cut down on some boilerplate from the test suite.

## New Functionality
- [x] Adds the following functions for equipartitioning the asset sets of maps and bundles:
    | Module | Function |
    | -- | -- |
    | `TokenMap` | `equipartitionAssets` |
    | `TokenBundle` | `equipartitionAssets` |
    
- [x] Adds an `assessBundleSize` parameter to `makeChange` and `performSelection`:
    ```hs
    assessBundleSize :: TokenBundle -> TokenBundleSizeAssessment
        
    data TokenBundleSizeAssessment
        = TokenBundleSizeWithinLimit
        -- ^ Indicates that the size of a token bundle does not exceed the maximum
        -- size that can be included in a transaction output.
        | TokenBundleSizeExceedsLimit
        -- ^ Indicates that the size of a token bundle exceeds the maximum size
        -- that can be included in a transaction output.
    ```
    
- [x] Adds function `splitBundlesWithExcessiveAssetCounts` to split change bundles with excessive asset counts, and calls it from `makeChange`.

- [x] Provides a definition of `assessTokenBundleSize` based on functionality exported by `Cardano.Api`.

## Testing

This PR adds:

- [x] Property tests for `equipartitionAssets` functions.  
- [x] Boundary unit tests for `performSelection` to show that change bundles are correctly split when they exceed the maximum size.
- [x] Unit tests for `Shelley.Compatibility.tokenBundleSizeAssessor` to verify that it gives sensible results for token bundles of various fixed sizes.
- [x] Property tests for `Shelley.Compatibility.tokenBundleSizeAssessor` to verify that it has the properties expected by `performSelection`.

jonathanknowles committed:
Move `unsafePartitionNatural` to module `Numeric.Util`.

jonathanknowles committed:
Move `equipartitionNatural` to module `Numeric.Util`.

jonathanknowles committed:
Move `equipartitionCoin` to module `Coin`.

jonathanknowles committed:
Move `equipartitionTokenQuantity` to module `TokenQuantity`.

jonathanknowles committed:
Move `inAscendingPartialOrder` to module `Numeric.Util`.

jonathanknowles committed:
Move `equipartitionTokenMap` to module `TokenMap`.

jonathanknowles committed:
Move `equipartitionTokenMapWithMaxQuantity` to module `TokenMap`.

jonathanknowles committed:
Move `equipartitionTokenBundleWithMaxQuantity` to module `TokenBundle`.

jonathanknowles committed:
Rename `equipartitionTokenBundlesWithMaxQuantity`.

Rename to `splitBundlesWithExcessiveTokenQuantities`.

Justification: this revised name more accurately captures the motivation
for splitting token bundles within the context of coin selection.

Internally, this function is still implemented in terms of
`TokenBundle.equipartitionQuantitiesWithUpperBound`.

jonathanknowles committed:
Add function `TokenMap.equipartitionAssets`.

jonathanknowles committed:
Add function `TokenBundle.equipartitionAssets`.

jonathanknowles committed:
Test `equipartitionQuantitiesWithUpperBound` only with positive upper bounds.

jonathanknowles committed:
Add function `splitBundleIfAssetCountExcessive`.

jonathanknowles committed:
Add function `splitBundlesWithExcessiveAssetCounts`.

jonathanknowles committed:
Adjust `makeChange` to take a single record value.

The argument list for `makeChange` is already very long, and is almost
identical to the fields of the `MakeChange` record data type.

Adjusting `makeChange` to accept a record type has the following advantages:

    1.  It's less cumbersome to add another argument to `makeChange`, because
        the compiler will tell us that a field is missing in all the required
        call sites (both in implementation code and in test code).

    2.  We can remove some duplication and boilerplate from the test code.

jonathanknowles committed:
Add `assessBundleSize` field to `MakeChangeCriteria`.

jonathanknowles committed:
Split up change maps with excessive asset counts in `makeChange`.

jonathanknowles committed:
Add `assessBundleSize` argument to `performSelection`.

jonathanknowles committed:
Add `bundleSizeAssessor` field to `BoundaryTestCriteria`.

jonathanknowles committed:
Add `BundleAssetCountUpperLimit` constructor to `BundleSizeAssessor`.

jonathanknowles committed:
Add boundary tests for `BundleAssetCountUpperLimit`.

jonathanknowles committed:
Use `mockAsset` and `mockAssetQuantity` to simplify boundary tests.

jonathanknowles committed:
Add token-bundle size assessment to the transaction layer.

This change adds an `assessTokenBundleSize` function to the transaction
layer, and to the `Shelley.Compatibility` module.

It also adds unit tests to verify that the `assessTokenBundleSize` gives
reasonable results for token bundles of fixed sizes.

jonathanknowles committed:
Consolidate the various definitions of `txOut{Min,Max}TokenQuantity`.

These were defined in multiple places.

Now we define them in one place: `Primitive.Types.Tx`.

jonathanknowles committed:
Move `genFixedSizeTokenBundle` to `TokenBundle.Gen`.

This will allow it to be reused by other test suites.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -881,22 +921,25 @@ makeChange minCoinFor requiredCost mExtraCoinSource inputBundles outputBundles
+            split b = b
+                & flip splitBundlesWithExcessiveAssetCounts
+                    (tokenBundleSizeExceedsLimit assessBundleSize)
+                & flip splitBundlesWithExcessiveTokenQuantities
```
In #2555 I ran into:
```
on\",\"message\":\"That's embarrassing. It looks like I've created an invalid transaction that could not be parsed by the node. Here's an error message that may help with debugging: HardForkApplyTxErrFromEra S (S (S (Z (WrapApplyTxErr {unwrapApplyTxErr = ApplyTxError [LedgerFailure (UtxowFailure (UtxoFailure (MaxTxSizeUTxO 4103 4096)))]}))))\"}")
```
while rebased to this branch.

I think this is due to `assignCoinsToChangeMaps` adding Coin values to the bundle _after_ we do the size check. 

The CBOR `uint` encodings have variable length. So a large `Coin` take up more space than a small `Coin`. ([cf](https://input-output-rnd.slack.com/archives/GBT05825V/p1615201247154700))


Using weights as Coin values when performing `splitBundlesWithExcessiveTokenQuantities` also seems bad then, since they affect the results.

And, I guess it would be <s>nice</s> pretty important to have `prop_performSelection` run with arbitrary size limits, and ensure all selections respect the `MaxTxSizeUTxO`

Anviking replied:
Wait, I might have misunderstood the error above (It's just the tx max size. I thought I raised it when testing), although I also believe the reasoning should also hold in theory 👀

Anviking replied:
This should be it:
```haskell
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=64)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=65)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=66)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=67) FAILED [1]
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=68) FAILED [2]
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=69)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=70)

Failures:

  /Users/anviking/IOHK/cardano-wallet/lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:747:13:
  1) API Specifications, SHELLEY_TRANSACTIONS, TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=67)
       expected: Status {statusCode = 202, statusMessage = "Accepted"}
        but got: Status {statusCode = 500, statusMessage = "Internal Server Error"}

       from the following response: Left (DecodeFailure "{\"code\":\"created_invalid_transaction\",\"message\":\"That's embarrassing. It looks like I've created an invalid transaction that could not be parsed by the node. Here's an error message that may help with debugging: HardForkApplyTxErrFromEra S (S (S (Z (WrapApplyTxErr {unwrapApplyTxErr = ApplyTxError [LedgerFailure (UtxowFailure (UtxoFailure (OutputTooBigUTxO [(Addr Mainnet (KeyHashObj (KeyHash \\\"f1946d14754ce4eb2a645aeff910bb279b88f0b3b69dc4f07cf28df5\\\")) (StakeRefBase (KeyHashObj (KeyHash \\\"509722367fde9ca9d9bdaee4c47353affff717df8f4fa82773ff828f\\\"))),Value 1997174061 (fromList [(PolicyID {policyID = ScriptHash \\\"4ff049585c4b3070563966370f5427d4a2f3588bce4146d57a93c7d3\\\"},fromList [(\\\"00000000000000000SeaHorse10\\\",1),(\\\"00000000000000000SeaHorse100\\\",1),(\\\"00000000000000000SeaHorse101\\\",1),(\\\"00000000000000000SeaHorse102\\\",1),(\\\"00000000000000000SeaHorse103\\\",1),(\\\"00000000000000000SeaHorse104\\\",1),(\\\"00000000000000000SeaHorse105\\\",1),(\\\"00000000000000000SeaHorse106\\\",1),(\\\"00000000000000000SeaHorse107\\\",1),(\\\"00000000000000000SeaHorse108\\\",1),(\\\"00000000000000000SeaHorse109\\\",1),(\\\"00000000000000000SeaHorse11\\\",1),(\\\"00000000000000000SeaHorse110\\\",1),(\\\"00000000000000000SeaHorse111\\\",1),(\\\"00000000000000000SeaHorse112\\\",1),(\\\"00000000000000000SeaHorse113\\\",1),(\\\"00000000000000000SeaHorse114\\\",1),(\\\"00000000000000000SeaHorse115\\\",1),(\\\"00000000000000000SeaHorse116\\\",1),(\\\"00000000000000000SeaHorse117\\\",1),(\\\"00000000000000000SeaHorse118\\\",1),(\\\"00000000000000000SeaHorse119\\\",1),(\\\"00000000000000000SeaHorse12\\\",1),(\\\"00000000000000000SeaHorse120\\\",1),(\\\"00000000000000000SeaHorse121\\\",1),(\\\"00000000000000000SeaHorse122\\\",1),(\\\"00000000000000000SeaHorse123\\\",1),(\\\"00000000000000000SeaHorse124\\\",1),(\\\"00000000000000000SeaHorse125\\\",1),(\\\"00000000000000000SeaHorse126\\\",1),(\\\"00000000000000000SeaHorse127\\\",1),(\\\"00000000000000000SeaHorse128\\\",1),(\\\"00000000000000000SeaHorse129\\\",1),(\\\"00000000000000000SeaHorse13\\\",1),(\\\"00000000000000000SeaHorse130\\\",1),(\\\"00000000000000000SeaHorse131\\\",1),(\\\"00000000000000000SeaHorse132\\\",1),(\\\"00000000000000000SeaHorse133\\\",1),(\\\"00000000000000000SeaHorse14\\\",1),(\\\"00000000000000000SeaHorse15\\\",1),(\\\"00000000000000000SeaHorse16\\\",1),(\\\"00000000000000000SeaHorse17\\\",1),(\\\"00000000000000000SeaHorse18\\\",1),(\\\"00000000000000000SeaHorse19\\\",1),(\\\"00000000000000000SeaHorse2\\\",1),(\\\"00000000000000000SeaHorse20\\\",1),(\\\"00000000000000000SeaHorse21\\\",1),(\\\"00000000000000000SeaHorse22\\\",1),(\\\"00000000000000000SeaHorse23\\\",1),(\\\"00000000000000000SeaHorse24\\\",1),(\\\"00000000000000000SeaHorse25\\\",1),(\\\"00000000000000000SeaHorse26\\\",1),(\\\"00000000000000000SeaHorse27\\\",1),(\\\"00000000000000000SeaHorse28\\\",1),(\\\"00000000000000000SeaHorse29\\\",1),(\\\"00000000000000000SeaHorse3\\\",1),(\\\"00000000000000000SeaHorse30\\\",1),(\\\"00000000000000000SeaHorse31\\\",1),(\\\"00000000000000000SeaHorse32\\\",1),(\\\"00000000000000000SeaHorse33\\\",1),(\\\"00000000000000000SeaHorse34\\\",1),(\\\"00000000000000000SeaHorse35\\\",1),(\\\"00000000000000000SeaHorse36\\\",1),(\\\"00000000000000000SeaHorse37\\\",1),(\\\"00000000000000000SeaHorse38\\\",1),(\\\"00000000000000000SeaHorse39\\\",1),(\\\"00000000000000000SeaHorse4\\\",1),(\\\"00000000000000000SeaHorse40\\\",1),(\\\"00000000000000000SeaHorse41\\\",1),(\\\"00000000000000000SeaHorse42\\\",1),(\\\"00000000000000000SeaHorse43\\\",1),(\\\"00000000000000000SeaHorse44\\\",1),(\\\"00000000000000000SeaHorse45\\\",1),(\\\"00000000000000000SeaHorse46\\\",1),(\\\"00000000000000000SeaHorse47\\\",1),(\\\"00000000000000000SeaHorse48\\\",1),(\\\"00000000000000000SeaHorse49\\\",1),(\\\"00000000000000000SeaHorse5\\\",1),(\\\"00000000000000000SeaHorse50\\\",1),(\\\"00000000000000000SeaHorse51\\\",1),(\\\"00000000000000000SeaHorse52\\\",1),(\\\"00000000000000000SeaHorse53\\\",1),(\\\"00000000000000000SeaHorse54\\\",1),(\\\"00000000000000000SeaHorse55\\\",1),(\\\"00000000000000000SeaHorse56\\\",1),(\\\"00000000000000000SeaHorse57\\\",1),(\\\"00000000000000000SeaHorse58\\\",1),(\\\"00000000000000000SeaHorse59\\\",1),(\\\"00000000000000000SeaHorse6\\\",1),(\\\"00000000000000000SeaHorse60\\\",1),(\\\"00000000000000000SeaHorse61\\\",1),(\\\"00000000000000000SeaHorse62\\\",1),(\\\"00000000000000000SeaHorse63\\\",1),(\\\"00000000000000000SeaHorse64\\\",1),(\\\"00000000000000000SeaHorse65\\\",1),(\\\"00000000000000000SeaHorse66\\\",1),(\\\"00000000000000000SeaHorse67\\\",1),(\\\"00000000000000000SeaHorse68\\\",1),(\\\"00000000000000000SeaHorse69\\\",1),(\\\"00000000000000000SeaHorse7\\\",1),(\\\"00000000000000000SeaHorse70\\\",1),(\\\"00000000000000000SeaHorse71\\\",1),(\\\"00000000000000000SeaHorse72\\\",1),(\\\"00000000000000000SeaHorse73\\\",1),(\\\"00000000000000000SeaHorse74\\\",1),(\\\"00000000000000000SeaHorse75\\\",1),(\\\"00000000000000000SeaHorse76\\\",1),(\\\"00000000000000000SeaHorse77\\\",1),(\\\"00000000000000000SeaHorse78\\\",1),(\\\"00000000000000000SeaHorse79\\\",1),(\\\"00000000000000000SeaHorse8\\\",1),(\\\"00000000000000000SeaHorse80\\\",1),(\\\"00000000000000000SeaHorse81\\\",1),(\\\"00000000000000000SeaHorse82\\\",1),(\\\"00000000000000000SeaHorse83\\\",1),(\\\"00000000000000000SeaHorse84\\\",1),(\\\"00000000000000000SeaHorse85\\\",1),(\\\"00000000000000000SeaHorse86\\\",1),(\\\"00000000000000000SeaHorse87\\\",1),(\\\"00000000000000000SeaHorse88\\\",1),(\\\"00000000000000000SeaHorse89\\\",1),(\\\"00000000000000000SeaHorse9\\\",1),(\\\"00000000000000000SeaHorse90\\\",1),(\\\"00000000000000000SeaHorse91\\\",1),(\\\"00000000000000000SeaHorse92\\\",1),(\\\"00000000000000000SeaHorse93\\\",1),(\\\"00000000000000000SeaHorse94\\\",1),(\\\"00000000000000000SeaHorse95\\\",1),(\\\"00000000000000000SeaHorse96\\\",1),(\\\"00000000000000000SeaHorse97\\\",1),(\\\"00000000000000000SeaHorse98\\\",1),(\\\"00000000000000000SeaHorse99\\\",1)])]))])))]}))))\"}")

  To rerun use: --match "/API Specifications/SHELLEY_TRANSACTIONS/TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=67)/"

  /Users/anviking/IOHK/cardano-wallet/lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:747:13:
  2) API Specifications, SHELLEY_TRANSACTIONS, TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=68)
       expected: Status {statusCode = 202, statusMessage = "Accepted"}
        but got: Status {statusCode = 500, statusMessage = "Internal Server Error"}

       from the following response: Left (DecodeFailure "{\"code\":\"created_invalid_transaction\",\"message\":\"That's embarrassing. It looks like I've created an invalid transaction that could not be parsed by the node. Here's an error message that may help with debugging: HardForkApplyTxErrFromEra S (S (S (Z (WrapApplyTxErr {unwrapApplyTxErr = ApplyTxError [LedgerFailure (UtxowFailure (UtxoFailure (OutputTooBigUTxO [(Addr Mainnet (KeyHashObj (KeyHash \\\"7ae290756d9b7b19e3c3620e45e13598d2998b9c8f7019d6aa3ec8b9\\\")) (StakeRefBase (KeyHashObj (KeyHash \\\"8defec29958c443bad8af705c2460baaa86621a544b761e867d728fc\\\"))),Value 1997159861 (fromList [(PolicyID {policyID = ScriptHash \\\"4ff049585c4b3070563966370f5427d4a2f3588bce4146d57a93c7d3\\\"},fromList [(\\\"00000000000000000SeaHorse10\\\",1),(\\\"00000000000000000SeaHorse100\\\",1),(\\\"00000000000000000SeaHorse101\\\",1),(\\\"00000000000000000SeaHorse102\\\",1),(\\\"00000000000000000SeaHorse103\\\",1),(\\\"00000000000000000SeaHorse104\\\",1),(\\\"00000000000000000SeaHorse105\\\",1),(\\\"00000000000000000SeaHorse106\\\",1),(\\\"00000000000000000SeaHorse107\\\",1),(\\\"00000000000000000SeaHorse108\\\",1),(\\\"00000000000000000SeaHorse109\\\",1),(\\\"00000000000000000SeaHorse11\\\",1),(\\\"00000000000000000SeaHorse110\\\",1),(\\\"00000000000000000SeaHorse111\\\",1),(\\\"00000000000000000SeaHorse112\\\",1),(\\\"00000000000000000SeaHorse113\\\",1),(\\\"00000000000000000SeaHorse114\\\",1),(\\\"00000000000000000SeaHorse115\\\",1),(\\\"00000000000000000SeaHorse116\\\",1),(\\\"00000000000000000SeaHorse117\\\",1),(\\\"00000000000000000SeaHorse118\\\",1),(\\\"00000000000000000SeaHorse119\\\",1),(\\\"00000000000000000SeaHorse12\\\",1),(\\\"00000000000000000SeaHorse120\\\",1),(\\\"00000000000000000SeaHorse121\\\",1),(\\\"00000000000000000SeaHorse122\\\",1),(\\\"00000000000000000SeaHorse123\\\",1),(\\\"00000000000000000SeaHorse124\\\",1),(\\\"00000000000000000SeaHorse125\\\",1),(\\\"00000000000000000SeaHorse126\\\",1),(\\\"00000000000000000SeaHorse127\\\",1),(\\\"00000000000000000SeaHorse128\\\",1),(\\\"00000000000000000SeaHorse129\\\",1),(\\\"00000000000000000SeaHorse13\\\",1),(\\\"00000000000000000SeaHorse130\\\",1),(\\\"00000000000000000SeaHorse131\\\",1),(\\\"00000000000000000SeaHorse132\\\",1),(\\\"00000000000000000SeaHorse133\\\",1),(\\\"00000000000000000SeaHorse134\\\",1),(\\\"00000000000000000SeaHorse135\\\",1),(\\\"00000000000000000SeaHorse14\\\",1),(\\\"00000000000000000SeaHorse15\\\",1),(\\\"00000000000000000SeaHorse16\\\",1),(\\\"00000000000000000SeaHorse17\\\",1),(\\\"00000000000000000SeaHorse18\\\",1),(\\\"00000000000000000SeaHorse19\\\",1),(\\\"00000000000000000SeaHorse2\\\",1),(\\\"00000000000000000SeaHorse20\\\",1),(\\\"00000000000000000SeaHorse21\\\",1),(\\\"00000000000000000SeaHorse22\\\",1),(\\\"00000000000000000SeaHorse23\\\",1),(\\\"00000000000000000SeaHorse24\\\",1),(\\\"00000000000000000SeaHorse25\\\",1),(\\\"00000000000000000SeaHorse26\\\",1),(\\\"00000000000000000SeaHorse27\\\",1),(\\\"00000000000000000SeaHorse28\\\",1),(\\\"00000000000000000SeaHorse29\\\",1),(\\\"00000000000000000SeaHorse3\\\",1),(\\\"00000000000000000SeaHorse30\\\",1),(\\\"00000000000000000SeaHorse31\\\",1),(\\\"00000000000000000SeaHorse32\\\",1),(\\\"00000000000000000SeaHorse33\\\",1),(\\\"00000000000000000SeaHorse34\\\",1),(\\\"00000000000000000SeaHorse35\\\",1),(\\\"00000000000000000SeaHorse36\\\",1),(\\\"00000000000000000SeaHorse37\\\",1),(\\\"00000000000000000SeaHorse38\\\",1),(\\\"00000000000000000SeaHorse39\\\",1),(\\\"00000000000000000SeaHorse4\\\",1),(\\\"00000000000000000SeaHorse40\\\",1),(\\\"00000000000000000SeaHorse41\\\",1),(\\\"00000000000000000SeaHorse42\\\",1),(\\\"00000000000000000SeaHorse43\\\",1),(\\\"00000000000000000SeaHorse44\\\",1),(\\\"00000000000000000SeaHorse45\\\",1),(\\\"00000000000000000SeaHorse46\\\",1),(\\\"00000000000000000SeaHorse47\\\",1),(\\\"00000000000000000SeaHorse48\\\",1),(\\\"00000000000000000SeaHorse49\\\",1),(\\\"00000000000000000SeaHorse5\\\",1),(\\\"00000000000000000SeaHorse50\\\",1),(\\\"00000000000000000SeaHorse51\\\",1),(\\\"00000000000000000SeaHorse52\\\",1),(\\\"00000000000000000SeaHorse53\\\",1),(\\\"00000000000000000SeaHorse54\\\",1),(\\\"00000000000000000SeaHorse55\\\",1),(\\\"00000000000000000SeaHorse56\\\",1),(\\\"00000000000000000SeaHorse57\\\",1),(\\\"00000000000000000SeaHorse58\\\",1),(\\\"00000000000000000SeaHorse59\\\",1),(\\\"00000000000000000SeaHorse6\\\",1),(\\\"00000000000000000SeaHorse60\\\",1),(\\\"00000000000000000SeaHorse61\\\",1),(\\\"00000000000000000SeaHorse62\\\",1),(\\\"00000000000000000SeaHorse63\\\",1),(\\\"00000000000000000SeaHorse64\\\",1),(\\\"00000000000000000SeaHorse65\\\",1),(\\\"00000000000000000SeaHorse66\\\",1),(\\\"00000000000000000SeaHorse67\\\",1),(\\\"00000000000000000SeaHorse68\\\",1),(\\\"00000000000000000SeaHorse69\\\",1),(\\\"00000000000000000SeaHorse7\\\",1),(\\\"00000000000000000SeaHorse70\\\",1),(\\\"00000000000000000SeaHorse71\\\",1),(\\\"00000000000000000SeaHorse72\\\",1),(\\\"00000000000000000SeaHorse73\\\",1),(\\\"00000000000000000SeaHorse74\\\",1),(\\\"00000000000000000SeaHorse75\\\",1),(\\\"00000000000000000SeaHorse76\\\",1),(\\\"00000000000000000SeaHorse77\\\",1),(\\\"00000000000000000SeaHorse78\\\",1),(\\\"00000000000000000SeaHorse79\\\",1),(\\\"00000000000000000SeaHorse8\\\",1),(\\\"00000000000000000SeaHorse80\\\",1),(\\\"00000000000000000SeaHorse81\\\",1),(\\\"00000000000000000SeaHorse82\\\",1),(\\\"00000000000000000SeaHorse83\\\",1),(\\\"00000000000000000SeaHorse84\\\",1),(\\\"00000000000000000SeaHorse85\\\",1),(\\\"00000000000000000SeaHorse86\\\",1),(\\\"00000000000000000SeaHorse87\\\",1),(\\\"00000000000000000SeaHorse88\\\",1),(\\\"00000000000000000SeaHorse89\\\",1),(\\\"00000000000000000SeaHorse9\\\",1),(\\\"00000000000000000SeaHorse90\\\",1),(\\\"00000000000000000SeaHorse91\\\",1),(\\\"00000000000000000SeaHorse92\\\",1),(\\\"00000000000000000SeaHorse93\\\",1),(\\\"00000000000000000SeaHorse94\\\",1),(\\\"00000000000000000SeaHorse95\\\",1),(\\\"00000000000000000SeaHorse96\\\",1),(\\\"00000000000000000SeaHorse97\\\",1),(\\\"00000000000000000SeaHorse98\\\",1),(\\\"00000000000000000SeaHorse99\\\",1)])]))])))]}))))\"}")

  To rerun use: --match "/API Specifications/SHELLEY_TRANSACTIONS/TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=68)/"

Randomized with seed 301508993

Finished in 41.1069 seconds
7 examples, 2 failures
```
First we succeed (no need to split change)
Then we fail (we should have split change, but we didn't)
Then we succeed (we should and do split the change)

Anviking replied:
And without this PR / rebased against master:
```
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=64)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=65)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=66)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=67) FAILED [1]
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=68) FAILED [2]
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=69) FAILED [3]
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=70) FAILED [4]
```
so it seems to make sense

jonathanknowles replied:
Hi @Anviking 

Many thanks for running these tests!

I've added a further commit 4c36492a6a9c5067ce25e6ad6aacc266cfa2960e which I believe fixes the problem you've highlighted.

I think your assessment here is totally correct:
> I think this is due to `assignCoinsToChangeMaps` adding `Coin` values to the bundle after we do the size check.

To fix this problem, we now perform size assessment based on what the size **_would be_** if the change map were paired with the **_maximum possible_** ada quantity:

```hs
assessBundleSizeWithMaxCoin
    :: TokenBundle -> TokenBundleSizeAssessment
assessBundleSizeWithMaxCoin =
    assessBundleSize . flip TokenBundle.setCoin (maxBound @Coin)
```
    
Of course, this approach has a (very small) chance of over-splitting a change map that would be marginally under the limit when combined its final ada quantity, but this would happen only very occasionally. I think this approach is certainly preferable to occasionally creating change bundles that are slightly over the limit, which would lead to transactions being rejected (as you have seen in testing).

Could you rebase your branch and let me know whether this enables the integration test to pass?

jonathanknowles replied:
Just for the record: I've rested this by rebasing @Anviking 's branch on top of this PR's branch, and the integration tests now pass: 🎉 

```hs
API Specifications
  SHELLEY_TRANSACTIONS
[cardano-wallet.integration:Notice:221] [2021-03-09 09:08:23.68 UTC] Configuring cluster in /tmp/test1945
[cardano-wallet.integration:Notice:228] [2021-03-09 09:08:25.97 UTC] Not registering any stake pools due to NO_POOLS=1. Some tests may fail.
[cardano-wallet.integration:Notice:255] [2021-03-09 09:08:26.15 UTC] Setting up faucet...
[cardano-wallet.integration:Notice:684] [2021-03-09 09:08:30.73 UTC] Wallet url: http://127.0.0.1:36019/
EKG url: none
Prometheus url: none
SMASH url: http://localhost:45375/

    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=64)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=65)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=66)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=67)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=68)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=69)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=70)

Finished in 44.8178 seconds
7 examples, 0 failures
```

jonathanknowles committed:
When assessing the size of a change map, use the maximum ada quantity.

When assessing the size of a change map to determine if it is
excessively large, we don't yet know how large the associated
ada quantity will be, since ada quantities are assigned at a
later stage (in 'assignCoinsToChangeMaps').

Therefore, we err on the side of caution, and assess the size
of a change map combined with the maximum possible ada quantity.

This means that when presented with a very large change map, we
have a small chance of splitting the map even if that map would
be within the limit when combined with its final ada quantity.

However, oversplitting a change map is preferable to creating
a bundle that is marginally over the limit, which would cause
the resultant transaction to be rejected.

In response to review feedback and discussion:

https://github.com/input-output-hk/cardano-wallet/pull/2552#discussion_r589335232

jonathanknowles committed:
Adjust `genFixedSizeTokenBundle` to generate boundary values more frequently.

This change adjusts the `genFixedSizeTokenBundle` generator so that it
has a much higher probability of generating boundary token quantities
(the minimum or maximum quantities that can appear in valid transaction
outputs).

As a consequence of this change, it was necessary to adjust the
following unit test to accept a range of expected sizes:

  `unit_assessTokenBundleSize_fixedSizeBundle`

The following unit test remains unaffected:

  `unit_computeMinimumAdaQuantity_fixedSizeBundle`

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1216,6 +1222,36 @@ toStakePoolDlgCert xpub (W.PoolId pid) =
+-- See: https://jira.iohk.io/projects/ADP/issues/ADP-779
+--
+maxTokenBundleSerializedLengthBytes :: Int
+maxTokenBundleSerializedLengthBytes = 4096
```
This should be `4000` (4 kB), and not `4096` (4 KiB).

Source: https://github.com/input-output-hk/cardano-ledger-specs/blob/5bccddb0ef0a1ae270b72fe0782a8ad680ed5852/shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Rules/Utxo.hs#L310

jonathanknowles replied:
Fixed in 82acfd0ca91e2859bc61d51cb51cff121b36fdee.

jonathanknowles committed:
Change `maxTokenBundleSerializedLengthBytes` to `4000`.

This constant should be 4000 bytes (4 kB), and not 4096 bytes (4 KiB).

Source:

https://github.com/input-output-hk/cardano-ledger-specs/blob/5bccddb0ef0a1ae270b72fe0782a8ad680ed5852/shelley-ma/impl/src/Cardano/Ledger/ShelleyMA/Rules/Utxo.hs#L310

Fixed in response to:

https://github.com/input-output-hk/cardano-wallet/pull/2552/files#r590109750

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -808,29 +842,11 @@ runSelectionStep lens s
-        -- ^ Token bundles of selected inputs.
-    -> NonEmpty TokenBundle
-        -- ^ Token bundles of original outputs.
+    :: MakeChangeCriteria (TokenMap -> Coin)
```
👍

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -1811,6 +1828,64 @@ unit_makeChangeForUserSpecifiedAsset =
+
+prop_splitBundleIfAssetCountExcessive_sum
+    :: Blind (Large TokenBundle) -> Positive Int -> Property
+prop_splitBundleIfAssetCountExcessive_sum
```
👍 

Wouldn't `"F.fold (splitBundleIfAssetCountExcessive b f) === b"` be a more readable `HSpec` it title? (More asking for properties in general in the future, not specifically here)

jonathanknowles replied:
> Wouldn't "F.fold (splitBundleIfAssetCountExcessive b f) === b" be a more readable HSpec it title? (More asking for properties in general in the future, not specifically here)

Possibly!

To be honest, I gave up trying put the definitions of properties in the titles. In simple cases like the one above, it works nicely, but in more complex cases, the titles can become excessively long, and the title just becomes a repetition of what is in the property anyway.

Another issue is that if someone changes the property definition, they may forget to change the title.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -802,10 +806,12 @@ runSelectionStep lens s
     { minCoinFor :: minCoinFor
       -- ^ A function that computes the minimum required ada quantity for a
       -- particular output.
+    , assessBundleSize :: assessBundleSize
```
Why abstract over this `assessBundleSize` type param? You already have `Coin` and `TokenBundle` types in other fields here.

Anviking replied:
Ah, right, you use it in the tests.

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -1292,6 +1298,161 @@ boundaryTest4 = BoundaryTestData
+          ]
+        )
+      , ( Coin 500_000
+        , [ (assetC, TokenQuantity 1)
```
👌

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs:
```diff
@@ -443,3 +448,23 @@ data TokenBundleSizeAssessment
+-- quantities are bounded by the size of a 'Word64'.
+--
+txOutMaxTokenQuantity :: TokenQuantity
+txOutMaxTokenQuantity = TokenQuantity $ fromIntegral $ maxBound @Word64
```
Ah. What's the cause for the "in practice ..."?

jonathanknowles replied:
> Ah. What's the cause for the "in practice ..."?

IIRC, the CBOR implementation requires this limit. This restriction isn't present in the original MA specification.

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/CompatibilitySpec.hs:
```diff
@@ -332,6 +357,72 @@ spec = do
+    -> Int
+    -- ^ Expected length (bytes)
+    -> Property
+unit_assessTokenBundleSize_fixedSizeBundle
```
👍

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs:
```diff
@@ -411,10 +408,12 @@ performSelection
         -- selection. This function must not depend on the magnitudes of
         -- individual asset quantities held within each change output.
+    -> (TokenBundle -> TokenBundleSizeAssessment)
+        -- ^ A function that assesses the size of a token bundle.
```
As we said in the call: we should describe the properties we expect from this function.

jonathanknowles replied:
Hi @Anviking 

I've added some documentation in commit 1d8f422f54aee089260c0a53c5d1fe287a8d9af7.

I'm also going to add some property tests for the CBOR implementation of `TokenBundleSizeAssessor`.

jonathanknowles replied:
Property tests added in commit 7f0059034d661b997933f1dc043f734d36a26c4e.

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobinSpec.hs:
```diff
@@ -1360,18 +1476,29 @@ linearCost SelectionSkeleton{inputsSkeleton, outputsSkeleton, changeSkeleton}
+        const TokenBundleSizeWithinLimit
+    BundleAssetCountUpperLimit upperLimit ->
+        \bundle ->
+            let assetCount = Set.size $ TokenBundle.getAssets bundle in
```
Also as previously said: 

We use `Set.size` to measure the size here in the unit tests, but the CBOR-encoded length in reality.

I fear there's no coverage for problems like what 4c36492 fixes — and that we may some day could end up generating too large bundles in a performSelection corner case.

But simply documenting what we expect in the argument to performSelection should probably be good enough.

jonathanknowles replied:
I've also added some further property tests in 7f0059034d661b997933f1dc043f734d36a26c4e.

Anviking reviewed (approved):
LGTM. The yet-to-be-added integration test now passes.

```haskell
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=64)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=65)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=66)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=67)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=68)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=69)
    TRANS_ASSETS_CREATE_02c - Send SeaHorses (nPerAddr=70)

Finished in 47.3657 seconds
```

Think https://github.com/input-output-hk/cardano-wallet/pull/2552#discussion_r590349490 should be fixed before merging.

jonathanknowles committed:
Add type `TokenBundleSizeAssessor`.

This type acts as a central point of reference for the expected properties
of token bundle size assessment functions.

This change also adjusts multiple function type signatures, replacing
`TokenBundle -> TokenBundleSizeAssessment` with `TokenBundleSizeAssessor`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/2552#discussion_r590349490

jonathanknowles committed:
Add property tests for `Shelley.Compatibility.tokenBundleSizeAssessor`.

In particular, we test that:

  * Enlarging a token bundle that is over the size limit should yield a token
    bundle that is still over the size limit.

  * Shrinking a token bundle that is within the size limit should yield a token
    bundle that is still within the size limit.

Commit:
Regenerate nix
